beluga_ros/tools/conversion_utils.py

```diff
@@ -0,0 +1,273 @@
+# Copyright 2024 Ekumen, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import h5py
+import yaml
+from typing import Dict, Optional
+from pathlib import Path
+import matplotlib.pyplot as plt
+from scipy.stats import multivariate_normal
+
+from dataclasses import dataclass
+
+
+@dataclass(frozen=True)
+class DiscreteCell:
+    x: int
+    y: int
+
+
+@dataclass(frozen=True)
+class NormalDistribution:
+    """
+    Wrapper around scipy's multivariate normal distribution implementation.
+
+    Allows for equality checks and typing hints.
+    """
+
+    mean: np.ndarray
+    covariance: np.ndarray
+
+    def get_scipy_representation(self):
+        """Get its scipy representation."""
+        return multivariate_normal(mean=self.mean, cov=self.covariance)
+
+    def __eq__(self, other: "NormalDistribution") -> bool:
+        """Element-wise equality."""
+        return np.allclose(self.mean, other.mean) and np.allclose(
+            self.covariance, other.covariance
+        )
+
+
+class NDTMap:
+    def __init__(self, resolution: float) -> None:
+        """Create a new NDT map with cell size of 'resolution'."""
+        self._resolution = resolution
+        self._grid: Dict[DiscreteCell, NormalDistribution] = {}
+
+    def add_distribution(self, cell: DiscreteCell, ndt: NormalDistribution):
+        """Add a new cell with its distribution."""
+        self._grid[cell] = ndt
+
+    def _get_plot_bounds(self):
+        min_x = float("inf")
+        min_y = float("inf")
+
+        max_x = -float("inf")
+        max_y = -float("inf")
+        for discrete_cell in self._grid.keys():
+            x = discrete_cell.x * self._resolution + self._resolution / 2
+            y = discrete_cell.y * self._resolution + self._resolution / 2
+            min_x = min(min_x, x)
+            max_x = max(max_x, x)
+
+            min_y = min(min_y, y)
+            max_y = max(max_y, y)
+
+        PADDING = 1  # [m]
+
+        min_x -= PADDING
+        max_x += PADDING
+        min_y -= PADDING
+        max_y += PADDING
+
+        step_x = (max_x - min_x) / 100.0
+        step_y = (max_y - min_y) / 100.0
+
+        xx, yy = np.mgrid[min_x:max_x:step_x, min_y:max_y:step_y]
+        return (xx, yy, np.dstack((xx, yy)))
+
+    def __eq__(self, other: "NDTMap") -> bool:
+        """Equality for two NDT maps."""
+        return (
+            np.allclose(self._resolution, other._resolution)
+            and self._grid == other._grid
+        )
```

Comment:
@serraramiro1 meta: all these equality comparisons come with an implicit tolerance. It is not really checking for equality. What about an `is_close` method instead, taking an optional tolerance?